File: source/src/main/java/org/rss/tools/mpl/global/ServiceProvider.java
package org.rss.tools.mpl.global;

import java.util.HashMap;
import java.util.Map;

/**
 * Responsible by providing an implementation, given an interface 
 * @author ricardo saturnino
 */
public final class ServiceProvider {

	private static final ReferenceWrapper REGISTER = new ReferenceWrapper();
	
	private ServiceProvider() {
	}

	public static <T> void register(Class<T> klazz, T instance) {
		REGISTER.put(klazz, instance);
	}

	public static <T> T get (Class<T> key) {
		T object = REGISTER.get(key);
		if (object == null) {
			throw new IllegalStateException("No instance was provieded for " + key);
		}
		return object;
	}

	final static class ReferenceWrapper {
		private final Map<Class<?>, Object> map = new HashMap<>();
		
		public <T> void put(Class<T> key, T value) {
			this.map.put(key, value);
		}
		
		/** Enforced by wrapped put */
		@SuppressWarnings("unchecked")
		public <T> T get(Class<T> key) {
			return (T) this.map.get(key);
		}
	}
}

File: source/src/main/kotlin/org/rss/tools/mpl/global/ServiceProvider.kt
package org.rss.tools.mpl.global

import java.util.*

/**
 * Responsible by providing an implementation, given an interface
 * @author ricardo saturnino
 */
object ServiceProvider {

    private val REGISTER = ReferenceWrapper()

    fun <T:Any> register(klazz: Class<T>, instance: T) {
        REGISTER.put(klazz, instance)
    }

    operator fun <T> get(key: Class<T>): T {
        return REGISTER[key] ?: throw IllegalStateException("No instance was provieded for $key")
    }

    internal class ReferenceWrapper {
        private val map = HashMap<Class<*>, Any>()

        fun <T:Any> put(key: Class<T>, value: T) {
            this.map[key] = value
        }

        /** Enforced by wrapped put  */
        operator fun <T> get(key: Class<T>): T? {
            return (this.map[key] as? T)
        }
    }
}
